main.lua

```diff
@@ -57,112 +60,128 @@ Ball = { -- single object
+    assert(type(y) == "number")
+    if rotation ~= nil then assert(type(rotation) == "number") else rotation = 0.0 end
 
     return {
-        spriteName = "brick_grey",
-        width = Brick.width,
-        height = Brick.height,
-        position = position,
+        spriteName = brickTemplate.spriteName,
+        width = brickTemplate.width,
+        height = brickTemplate.height,
+        position = Vector.new(x, y),
         rotation = rotation,
     }
 end
 
-function ResetGame()
-    -- setup paddle
-    Paddle.position = Vector.new((Window.width - Paddle.width) / 2, Window.height - 100)
-    Paddle.velocity = Vector.new()
-    Paddle.acceleration = Vector.new()
+local function resetBall()
+    ball.position:setXY(window.width / 2, paddle.position.y - (ball.height + 8))
+    ball.velocity:setXY(0, 1)
+end
+
+local function resetPaddle()
+    paddle.position:setXY(window.width / 2, window.height - 100)
+    paddle.velocity:setXY(0, 0)
+    paddle.acceleration:setXY(0, 0)
 
-    -- check for keys pressed before game setup
-    if love.keyboard.isDown(KeyEnum.LEFT) then
-        Paddle.acceleration:transform(LeftVector)
-    end
-    if love.keyboard.isDown(KeyEnum.RIGHT) then
-        Paddle.acceleration:transform(RightVector)
-    end
+    -- check for keys pressed and not released before resetting
+    if love.keyboard.isDown(KeyEnum.LEFT) then paddle.acceleration:transform(left) end
+    if love.keyboard.isDown(KeyEnum.RIGHT) then paddle.acceleration:transform(right) end
+end
+
+local function resetGame()
+    -- setup global state
+    livesLeft = 3
 
-    -- setup ball
-    Ball.position = Vector.new((Window.width - Ball.width) / 2, (Window.height - Ball.height) / 2)
-    Ball.velocity = Vector.fromAngle(math.rad(-90 + math.random(-30, 30)))
+    -- setup paddle and ball
+    resetPaddle()
+    resetBall()
 
     -- setup bricks
-    local centerX = Window.width / 2
-    local centerY = Window.height / 2 - 100
-    Bricks = {}
-    for dx = -3, 4 do
+    local centerX = window.width / 2
+    local centerY = window.height / 2 - 100
+
+    bricks = {}
+    for dx = -3, 3 do
         for dy = -3, 0 do
-            local x = centerX + dx * Brick.width
-            local y = centerY + dy * Brick.height
-            local r = (dx - 0.5) / 3.5 * math.rad(22.5)
+            local x = centerX + dx * brickTemplate.width
+            local y = centerY + dy * brickTemplate.height
+            local r = dx / 3.5 * math.rad(22.5)
 
-            table.insert(Bricks, CreateBrick(x, y, r))
+            table.insert(bricks, createBrick(x, y, r))
         end
     end
 end
 
-function UpdateBall(dt)
+local function updateBall(dt)
     -- calculate ball position after this update
-    local dBallPosition = Ball.velocity:copy() * Ball.dVelocity * dt
-    local newBallPosition = Ball.position + dBallPosition
+    local dBallPosition = ball.velocity:copy() * ball.dVelocity * dt
+    local newBallPosition = ball.position + dBallPosition
```

Comment:
Doesn't `__mul` return a new instance already? Do we need to call copy?

Reply:
`dBallPosition` is a movement vector relative to current position, while `newBallPosition` is an exact position after the movement

Reply:
But yes, `__mul` returns a new instance

Reply:
Oh... didn't undestand. Good point, `copy()` is redundant

Reply:
Fixed in 6fb4e9591fbb6066c6f76274e62d1785c2009b2f